fix(smart-apply): error when applying code to an empty file

ichim-david opened the pull request:
Refs CODY-5396 
Fixed bug where inserting any code within an empty file triggered an error
which doesn't abort the smart apply action, leaving the user to  think it's super slow.

This started after the work here https://github.com/sourcegraph/cody/pull/6768
Because `startStreamingEditTask` calls fixup `startDecorator` 
https://github.com/sourcegraph/cody/pull/6768/files#diff-aab0c624f666dd0c22f4415b5d9cf6e2b1a9d3ceb14526e2714896beade885dfR291

And we bump the end line by one even when the file is empty
https://github.com/sourcegraph/cody/blob/main/vscode/src/non-stop/FixupTask.ts#L155
The `trimEmptyLinesFromRange` ended up with an index higher than available which 
crashed https://github.com/sourcegraph/cody/pull/7439/files#diff-e0078048d45c5a1b5f5eea5af56c3919ecf75b0fe610dfee87578e87d0944ae0R87

Also used the correct model name to ensure we get 15k input instead of the default
fallback of 7k in case we have a large file that requires edits.

## Test plan
Apply code in an empty file and observe it working

umpox commented on lib/shared/src/models/modelsService.ts:
```diff
@@ -591,9 +591,14 @@ export class ModelsService {
 
-    public getModelByID(modelID: string): Model | undefined {
+    public getModelByID(modelID: string, models = this.models): Model | undefined {
         // TODO(sqs)#observe: remove synchronous access here, return an Observable<Model|undefined> instead
-        return this.models.find(m => m.id === modelID)
+        // split on :: or / to get the model id
+        // e.g. "openai::v1::gpt-4o-mini" or "openai/gpt-4o-mini"
+        const modelParameters = modelID.split(/[:\/]{2}|\//)
+        const modelId = modelParameters.at(-1) || ''
+
+        return models.find(m => m.id.endsWith(modelId))
     }
```
Can we consider breaking this logic out into a different PR? It's worrying to me that it seems this logic was broken before, do we have the incorrect context window for other models too?

It'd also be useful to get some tests as splitting via a regex is going to be harder to maintain for future models than just matching an id

ichim-david commented on vscode/src/non-stop/decorations/compute-decorations.test.ts:
```diff
@@ -103,4 +103,28 @@ describe('computeOngoingDecorations', () => {
         expect(decorations?.unvisitedLines).toStrictEqual([])
     })
+
+    it('inserts content into empty document', () => {
```
I'm not so thrilled with this test because it's hard to understand the appliedDecorations and the lines removed or unvisited lines.
I had to tinker from the LLM test generation to creating something that passed but the function itself
is a bit funky.

A more valid test is what I wrote as an e2e test where you can actually see a smart apply insert
https://github.com/sourcegraph/cody/pull/7439/files#diff-1b6cb520ccf70c987b273ee5d3d1e297389ca7ae5150ea0379a1703efa9f1828R112

ichim-david commented on vscode/src/services/utils/codeblock-action-tracker.ts:
```diff
@@ -173,7 +173,7 @@ function getSmartApplyExperimentModel(
 async function getSmartApplyModel(authStatus: AuthStatus): Promise<EditModel | undefined> {
     if (isDotCom(authStatus) || isS2(authStatus)) {
-        const defaultModel: EditModel = 'anthropic/claude-3-5-sonnet-20240620'
+        const defaultModel: EditModel = 'anthropic::2024-10-22::claude-3-5-sonnet-latest'
```
Th is is the id of the model, false alarm

ichim-david commented:
@umpox this is ready for review, finally.

I added a check for any potential startStreamingEdit errors as you will see.
I don't know if we should also display the vscode error popup or not,
you do get the Applying Edit error but it shows up on the editor instead of the smart apply chat panel section.
![error-vscode-debug-and-edit-error](https://github.com/user-attachments/assets/e346163a-27d5-4f70-8b93-cb6e93ac178d)

You can actually click on the `Applying error` and you get a popup with the error message but I didn't know that
it was possible, as such I kind of feel that it's a good thing to point to an error message in the console.
![error-vscode-debug-and-edit-error-clicking-on-applying-edit-failed](https://github.com/user-attachments/assets/346e9f73-6b4a-4a1b-89a8-1edc8b272729)

Let me know if you have any further suggestions (I have removed the regex check for the modelID for another pull request).

Check the tests I wrote as well, especially Smart Apply e2e one
https://github.com/sourcegraph/cody/pull/7439/files#diff-1b6cb520ccf70c987b273ee5d3d1e297389ca7ae5150ea0379a1703efa9f1828R112

hitesh-1997 commented on lib/shared/src/models/modelsService.ts:
```diff
@@ -624,7 +624,7 @@ export class ModelsService {
     }
 
-    public getModelByID(modelID: string): Model | undefined {
+    public getModelByID(modelID: string, models = this.models): Model | undefined {
```
nit: This refactor seems unnecessary. If preferred we can keep as is.

ichim-david replied:
Removed the changes for another pull request

ichim-david commented on vscode/test/fixtures/workspace/smartApply.py:
@hitesh-1997 this is part of the e2e test to apply to an empty file https://github.com/sourcegraph/cody/pull/7439/files#diff-1b6cb520ccf70c987b273ee5d3d1e297389ca7ae5150ea0379a1703efa9f1828R112

hitesh-1997 commented on vscode/src/edit/edit-manager.ts:
```diff
@@ -306,9 +307,23 @@ export class EditManager implements vscode.Disposable {
+            // If there's an error, abort the task by sending an error event
+            this.options.fixupController.error(task.id, errorObject)
+
+            logError('startStreamingEditTask error:', errorObject.message)
```
Why do we need special handling here now ? Won't error always propogate to upstream. Did you observe any instances where errors are thrown ? I am assuming there would be upstream services handling the errors, so  can this `vscode.window.showErrorMessage ` lead to duplication.
Adding @umpox who has more idea about the application code.

ichim-david replied:
@hitesh-1997 not in the case of the `startStreamingEditTask` task. When I started with this ticket there was a logic bug within `trimEmptyLinesFromRange` https://github.com/sourcegraph/cody/blob/ichimdav/apply_markdown/vscode/src/non-stop/decorations/compute-decorations.ts#L82

Which gave the error that you see in the comment I left here
https://github.com/sourcegraph/cody/pull/7439#issuecomment-2785509278

Before my code it would simply fail silently but the user never new and the smart apply was just spinning without it triggering an error that allows you to see the error.

I have mentioned in the comment above that it might be excessive to add the vscode `showErrorMessage` which is why I was asking if we should have it or not.

hitesh-1997 commented on vscode/src/edit/prompt/smart-apply/selection/default.ts:
```diff
@@ -63,6 +63,11 @@ export class DefaultSelectionProvider implements SmartApplySelectionProvider {
         const documentText = PromptString.fromDocumentText(document, documentRange)
+        // If the document is empty, we should insert the code without call to LLM
+        // to decide what to choose as we do with the custom model that returns early
+        if (!documentText.toString()) {
```
We don't use `default.ts` for the new smart apply model which has majority of the traffic. Do we need to add this to other selection providers.

ichim-david replied:
@hitesh-1997 the new smart-apply has it's own early return in check for entire file
https://github.com/sourcegraph/cody/blob/ichimdav/apply_markdown/vscode/src/edit/prompt/smart-apply/selection/custom-model.ts#L61

hitesh-1997 commented on vscode/src/edit/prompt/smart-apply/selection/default.ts:
```diff
@@ -63,6 +63,11 @@ export class DefaultSelectionProvider implements SmartApplySelectionProvider {
+        // If the document is empty, we should insert the code without call to LLM
+        // to decide what to choose as we do with the custom model that returns early
+        if (!documentText.toString()) {
+            return 'INSERT'
```
can we define this as constant instead of hardcoding here

ichim-david replied:
@hitesh-1997 what do you think about this commit? https://github.com/sourcegraph/cody/pull/7439/commits/7d60d2d0e9b9509dd02b2a9f3c469dd075fff4c6
Moved both ENTIRE_FILE and INSERT into constant values

hitesh-1997 reviewed (changes requested):
minor comments

abeatrix commented on vscode/src/edit/edit-manager.ts:
```diff
@@ -306,9 +307,23 @@ export class EditManager implements vscode.Disposable {
+            this.options.fixupController.error(task.id, errorObject)
+
+            logError('EditManager', 'startStreamingEditTask', errorObject)
+            void vscode.window.showErrorMessage(
+                `Error performing the ${task.intent} task. Check the logs for more details.`
+            )
```
```suggestion
```

I think we should avoid using vscode.window.showErrorMessage here as it's not supported across editors (visual studio and eclipse dont have it implemented afaik, plus we should probably handle all error the same way so `this.options.fixupController.error` might probably the better place if we do

ichim-david commented:
@hitesh-1997 when you have time please let me know if my comments and changes have not addressed your concerns so that I can get the stamp of approval for this work.
